packages/ruleset/src/functions/no-superfluous-allof.js

```diff
@@ -0,0 +1,37 @@
+/**
+ * Copyright 2017 - 2024 IBM Corporation.
+ * SPDX-License-Identifier: Apache2.0
+ */
+
+const { validateSubschemas } = require('@ibm-cloud/openapi-ruleset-utilities');
+const { LoggerFactory } = require('../utils');
+
+let ruleId;
+let logger;
+
+module.exports = function (schema, _opts, context) {
+  if (!logger) {
+    ruleId = context.rule.name;
+    logger = LoggerFactory.getInstance().getLogger(ruleId);
+  }
+
+  return validateSubschemas(schema, context.path, checkForSuperfluousAllOf);
+};
+
+function checkForSuperfluousAllOf(schema, path) {
+  // We're interested only in schemas that contain ONLY a single-element allOf list.
+  if (
+    Array.isArray(schema.allOf) &&
+    schema.allOf.length === 1 &&
+    Object.keys(schema).length === 1
+  ) {
+    return [
+      {
+        message: 'Avoid schemas with a superfluous allOf attribute',
```

Comment:
Picky, optional suggestion to slightly reword since it's not the _schema_ that needs to be avoided but the `allOf`.

```suggestion
        message: 'Avoid using superfluous allOf attributes in schemas',
```

Reply:
After thinking about this a little, I think the error message contained in the rule definition itself (`Avoid schemas containing only a single-element allOf`) might capture the essence of what we're trying to say.  It's not ONLY the single-element allOf that is the "problem"... it's the fact that a schema essentially contains nothing else BUT a single-element allOf. A single-element allOf is perfectly fine within a ref-sibling pattern, for example.  The reason that an allOf might be "superfluous" is that it contains only a single element AND the containing schema doesn't contain anything else.

Having said all that, I think an argument can be made that the path that we report in the error entry should probably reflect the schema that contains the allOf, and not the allOf itself. I.e. it's the otherwise empty schema that renders the single-element allOf as "superfluous".

So I'll make these changes and then we can take another look at the PR and see if that looks ok.